### PR TITLE
Add quality profiles for Weedo X40

### PR DESCRIPTION
**Pull request by x40-Community**
Quality profiles for Weedo X40 printer

### CODE REVIEW

File: resources/intent/weedo_x40/Extrudr/weedo_x40_0.4_Extrudr_GreenTECPro_Anthracit_Extra_Fine_Engineering.inst.cfg [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+version = 4
+name = Accurate
+definition = weedo_x40
+
+[metadata]
+setting_version = 17
+type = intent
+quality_type = extra fine
+intent_category = engineering
+material = Extrudr_GreenTECPro_Black_175
```
Comment (nallath) [inline resources/intent/weedo_x40/Extrudr/weedo_x40_0.4_Extrudr_GreenTECPro_Anthracit_Extra_Fine_Engineering.inst.cfg:11]: So the engineering intent is only there for the black material? That seems a bit weird. 

Ah. I see the issue now; the original profiles were also only there for black material, which is also weird...

File: resources/intent/weedo_x40/Extrudr/weedo_x40_0.4_Extrudr_GreenTECPro_Anthracit_Extra_Fine_Engineering.inst.cfg [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+speed_equalize_flow_enabled = True
+acceleration_enabled = True
+acceleration_print = 400
+acceleration_travel = 800
+acceleration_travel_layer_0 = =acceleration_travel / 2
+
+jerk_enabled = True
+line_width = =0.02 + 0.7 * machine_nozzle_size
+
+speed_print = 30
```
Comment (nallath) [inline resources/intent/weedo_x40/Extrudr/weedo_x40_0.4_Extrudr_GreenTECPro_Anthracit_Extra_Fine_Engineering.inst.cfg:27]: No need to re set the speed. The quality profile already sets it to 30

File: resources/intent/weedo_x40/Extrudr/weedo_x40_0.4_Extrudr_GreenTECPro_Anthracit_Extra_Fine_Engineering.inst.cfg [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+top_bottom_thickness = =wall_thickness
```
Comment (nallath) [inline resources/intent/weedo_x40/Extrudr/weedo_x40_0.4_Extrudr_GreenTECPro_Anthracit_Extra_Fine_Engineering.inst.cfg:28]: It seems that you're setting this for every intent. Shouldn't it be in the quality instead? (or should it be set like this for all intents except for default?)

File: resources/intent/weedo_x40/Extrudr/weedo_x40_0.4_Extrudr_GreenTECPro_Anthracit_Fine_Engineering.inst.cfg [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+variant = 0.4mm Nozzle
+
+[values]
+infill_sparse_density = 30
+outer_inset_first = True
+infill_line_width = =line_width
+acceleration_enabled = True
+jerk_enabled = True
+top_bottom_thickness = =wall_thickness
+wall_line_width_x = =line_width
```
Comment (nallath) [inline resources/intent/weedo_x40/Extrudr/weedo_x40_0.4_Extrudr_GreenTECPro_Anthracit_Fine_Engineering.inst.cfg:21]: This is already the default in fdmprinter. No need to overide it again

File: resources/intent/weedo_x40/Extrudr/weedo_x40_0.4_Extrudr_GreenTECPro_Red_Draft_Draft.inst.cfg [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+version = 4
+name = Quick
+definition = weedo_x40
+
+[metadata]
+setting_version = 17
+type = intent
+quality_type = draft
+intent_category = quick
+material = Extrudr_GreenTECPro_Black_175
```
Comment (nallath) [inline resources/intent/weedo_x40/Extrudr/weedo_x40_0.4_Extrudr_GreenTECPro_Red_Draft_Draft.inst.cfg:11]: The ID of the profile says _greentecpro_red, but the material is greentecpro_black

I think it's better to just add greentecpro as a generic material and let all of them reference to that

File: resources/quality/weedo_x40/Extrudr/weedo_x40_0.4_Extrudr_GreenTECPro_Anthracit_Draft.inst.cfg [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+layer_height_0 = =round(layer_height * 1.25, 2)
+material_print_temperature_layer_0 = =material_print_temperature + 10
+material_initial_print_temperature = =material_print_temperature - 5
+material_final_print_temperature = =material_print_temperature - 5
+material_bed_temperature = =default_material_bed_temperature
+material_bed_temperature_layer_0= =material_bed_temperature + 10
+material_flow_layer_0 = =material_flow + 4
+skirt_brim_material_flow= =material_flow + 4
+material_standby_temperature = 100
+support_z_distance = 0.2
```
Comment (nallath) [inline resources/quality/weedo_x40/Extrudr/weedo_x40_0.4_Extrudr_GreenTECPro_Anthracit_Draft.inst.cfg:25]: Why override this again? The definition sets it to  "layer_height if layer_height >= 0.16 else layer_height*2"

**Review by nallath — CHANGES_REQUESTED** [review]
I've checked some of the settings at random and found quite a few cases where settings were overridden without it needing to be like that. 

Please double check if the changes that you make are actually needed. This will make it a lot easier to maintain the profiles in the future.

**Comment by Ghostkeeper** [comment]
There are conflicts still, but by your attempts to resolve those conflicts I think we'll know how to resolve those.

**Comment by Ghostkeeper** [comment]
To be merged together with this PR: https://github.com/Ultimaker/fdm_materials/pull/183

**Comment by Ghostkeeper** [comment]
This PR has quite a big maintenance problem, and I'm not really sure how to deal with it here. It's a lot of work to fix something like this or even to point out where the problem specifically lies. The root of the problem is performance here: This PR adds 1100 profiles, but only 24 unique sets of settings. Most of them are duplicates. This is a result of two issues that need to be addressed:

Duplicate profiles for colours
----
This is one set of profiles which are all duplicates:
```
weedo_x40_0.4_Material4Print_PETG_Black_Fine_Engineering.inst.cfg
weedo_x40_0.4_Material4Print_PETG_Black_Normal_Engineering.inst.cfg
weedo_x40_0.4_Material4Print_PETG_Brightgreen_Fine_Engineering.inst.cfg
weedo_x40_0.4_Material4Print_PETG_Brightgreen_Normal_Engineering.inst.cfg
weedo_x40_0.4_Material4Print_PETG_Brightorange_Fine_Engineering.inst.cfg
weedo_x40_0.4_Material4Print_PETG_Brightorange_Normal_Engineering.inst.cfg
weedo_x40_0.4_Material4Print_PETG_Brightred_Fine_Engineering.inst.cfg
weedo_x40_0.4_Material4Print_PETG_Brightred_Normal_Engineering.inst.cfg
weedo_x40_0.4_Material4Print_PETG_Dustygrey_Fine_Engineering.inst.cfg
weedo_x40_0.4_Material4Print_PETG_Dustygrey_Normal_Engineering.inst.cfg
weedo_x40_0.4_Material4Print_PETG_Greyaluminium_Fine_Engineering.inst.cfg
weedo_x40_0.4_Material4Print_PETG_Greyaluminium_Normal_Engineering.inst.cfg
weedo_x40_0.4_Material4Print_PETG_Irongrey_Fine_Engineering.inst.cfg
weedo_x40_0.4_Material4Print_PETG_Irongrey_Normal_Engineering.inst.cfg
weedo_x40_0.4_Material4Print_PETG_Purered_Fine_Engineering.inst.cfg
weedo_x40_0.4_Material4Print_PETG_Purered_Normal_Engineering.inst.cfg
weedo_x40_0.4_Material4Print_PETG_Signalblue_Fine_Engineering.inst.cfg
weedo_x40_0.4_Material4Print_PETG_Signalblue_Normal_Engineering.inst.cfg
weedo_x40_0.4_Material4Print_PETG_Skyblue_Fine_Engineering.inst.cfg
weedo_x40_0.4_Material4Print_PETG_Skyblue_Normal_Engineering.inst.cfg
weedo_x40_0.4_Material4Print_PETG_Trafficyellow_Fine_Engineering.inst.cfg
weedo_x40_0.4_Material4Print_PETG_Trafficyellow_Normal_Engineering.inst.cfg
weedo_x40_0.4_Material4Print_PETG_Transparentblack_Fine_Engineering.inst.cfg
weedo_x40_0.4_Material4Print_PETG_Transparentblack_Normal_Engineering.inst.cfg
weedo_x40_0.4_Material4Print_PETG_Transparentblue_Fine_Engineering.inst.cfg
weedo_x40_0.4_Material4Print_PETG_Transparentblue_Normal_Engineering.inst.cfg
weedo_x40_0.4_Material4Print_PETG_Transparentgreen_Fine_Engineering.inst.cfg
weedo_x40_0.4_Material4Print_PETG_Transparentgreen_Normal_Engineering.inst.cfg
weedo_x40_0.4_Material4Print_PETG_Transparentred_Normal_Engineering.inst.cfg
weedo_x40_0.4_Material4Print_PETG_Transparentyellow_Fine_Engineering.inst.cfg
weedo_x40_0.4_Material4Print_PETG_Transparent_Fine_Engineering.inst.cfg
weedo_x40_0.4_Material4Print_PETG_Transparent_Normal_Engineering.inst.cfg
weedo_x40_0.4_Material4Print_PETG_White_Fine_Engineering.inst.cfg
weedo_x40_0.4_Material4Print_PETG_White_Normal_Engineering.inst.cfg
weedo_x40_0.4_Material4Print_PLA_Anthracite_Fine_Engineering.inst.cfg
weedo_x40_0.4_Material4Print_PLA_Anthracite_Normal_Engineering.inst.cfg
weedo_x40_0.4_Material4Print_PLA_Black_Fine_Engineering.inst.cfg
weedo_x40_0.4_Material4Print_PLA_Black_Normal_Engineering.inst.cfg
weedo_x40_0.4_Material4Print_PLA_Blue_Fine_Engineering.inst.cfg
weedo_x40_0.4_Material4Print_PLA_Blue_Normal_Engineering.inst.cfg
weedo_x40_0.4_Material4Print_PLA_Brightorange_Normal_Engineering.inst.cfg
weedo_x40_0.4_Material4Print_PLA_Brightred_Fine_Engineering.inst.cfg
weedo_x40_0.4_Material4Print_PLA_Brightred_Normal_Engineering.inst.cfg
weedo_x40_0.4_Material4Print_PLA_Brown_Fine_Engineering.inst.cfg
weedo_x40_0.4_Material4Print_PLA_Brown_Normal_Engineering.inst.cfg
weedo_x40_0.4_Material4Print_PLA_Dustygrey_Fine_Engineering.inst.cfg
weedo_x40_0.4_Material4Print_PLA_Dustygrey_Normal_Engineering.inst.cfg
weedo_x40_0.4_Material4Print_PLA_Green_Fine_Engineering.inst.cfg
weedo_x40_0.4_Material4Print_PETG_Transparentred_Fine_Engineering.inst.cfg
weedo_x40_0.4_Material4Print_PETG_Transparentyellow_Normal_Engineering.inst.cfg
weedo_x40_0.4_Material4Print_PLA_Brightorange_Fine_Engineering.inst.cfg
weedo_x40_0.4_Material4Print_PLA_Greyaluminium_Fine_Engineering.inst.cfg
weedo_x40_0.4_Material4Print_PLA_Greyaluminium_Normal_Engineering.inst.cfg
weedo_x40_0.4_Material4Print_PLA_Irongrey_Fine_Engineering.inst.cfg
weedo_x40_0.4_Material4Print_PLA_Irongrey_Normal_Engineering.inst.cfg
weedo_x40_0.4_Material4Print_PLA_Ivory_Fine_Engineering.inst.cfg
weedo_x40_0.4_Material4Print_PLA_Ivory_Normal_Engineering.inst.cfg
weedo_x40_0.4_Material4Print_PLA_Magenta_Fine_Engineering.inst.cfg
weedo_x40_0.4_Material4Print_PLA_Magenta_Normal_Engineering.inst.cfg
weedo_x40_0.4_Material4Print_PLA_Nature_Fine_Engineering.inst.cfg
weedo_x40_0.4_Material4Print_PLA_Nature_Normal_Engineering.inst.cfg
weedo_x40_0.4_Material4Print_PLA_Pearlblue_Normal_Engineering.inst.cfg
weedo_x40_0.4_Material4Print_PLA_Perlgold_Fine_Engineering.inst.cfg
weedo_x40_0.4_Material4Print_PLA_Perlgold_Normal_Engineering.inst.cfg
weedo_x40_0.4_Material4Print_PLA_Purered_Fine_Engineering.inst.cfg
weedo_x40_0.4_Material4Print_PLA_Purered_Normal_Engineering.inst.cfg
weedo_x40_0.4_Material4Print_PLA_Purple_Fine_Engineering.inst.cfg
weedo_x40_0.4_Material4Print_PLA_Purple_Normal_Engineering.inst.cfg
weedo_x40_0.4_Material4Print_PLA_Signalblue_Fine_Engineering.inst.cfg
weedo_x40_0.4_Material4Print_PLA_Trafficyellow_Fine_Engineering.inst.cfg
weedo_x40_0.4_Material4Print_PLA_Trafficyellow_Normal_Engineering.inst.cfg
weedo_x40_0.4_Material4Print_PLA_White_Fine_Engineering.inst.cfg
weedo_x40_0.4_Material4Print_PLA_White_Normal_Engineering.inst.cfg
weedo_x40_0.4_Material4Print_PLA_Yellowgreen_Fine_Engineering.inst.cfg
weedo_x40_0.4_Material4Print_PLA_Yellowgreen_Normal_Engineering.inst.cfg
weedo_x40_0.4_Material4Print_PLA_Green_Normal_Engineering.inst.cfg
weedo_x40_0.4_Material4Print_PLA_Pearlblue_Fine_Engineering.inst.cfg
weedo_x40_0.4_Material4Print_PLA_Signalblue_Normal_Engineering.inst.cfg
```
All with the following overrides:
```
acceleration_enabled=True
infill_sparse_density=30
jerk_enabled=True
outer_inset_first=True
top_bottom_thickness==wall_thickness
wall_thickness==line_width * 3
xy_offset==-layer_height * 0.2
```

This indicates that there are unique profiles for every colour of filament. However this is unnecessary. Profiles match on materials with a three-tier system:
* If there are any quality/intent profiles specific to the current material the list of those quality profiles is shown.
* Otherwise, it will show the list of all quality/intent profiles specific to a material of the same *material type* as the current material.
* If there are none of those, it will use an empty quality/intent profile.

As such, you don't need to make a unique profile for each colour, only a unique profile for each material type. You can still make a profile for a specific colour if that material needs different settings due to its pigment or something.

Improper material types
----
This is a second set of duplicate profiles:
```
weedo_x40_0.4_Extrudr_GreenTECPro_Anthracit_Normal_Visual.inst.cfg
weedo_x40_0.4_Extrudr_GreenTECPro_Black_Normal_Visual.inst.cfg
weedo_x40_0.4_Extrudr_GreenTECPro_Blue_Normal_Visual.inst.cfg
weedo_x40_0.4_Extrudr_GreenTECPro_Nature_Normal_Visual.inst.cfg
weedo_x40_0.4_Extrudr_GreenTECPro_Red_Normal_Visual.inst.cfg
weedo_x40_0.4_Extrudr_GreenTECPro_Silver_Normal_Visual.inst.cfg
weedo_x40_0.4_Extrudr_GreenTECPro_White_Normal_Visual.inst.cfg
weedo_x40_0.4_Material4Print_PETG_Transparentblack_Normal_Visual.inst.cfg
weedo_x40_0.4_Material4Print_PETG_Black_Normal_Visual.inst.cfg
weedo_x40_0.4_Material4Print_PETG_Brightgreen_Normal_Visual.inst.cfg
weedo_x40_0.4_Material4Print_PETG_Brightorange_Normal_Visual.inst.cfg
weedo_x40_0.4_Material4Print_PETG_Dustygrey_Normal_Visual.inst.cfg
weedo_x40_0.4_Material4Print_PETG_Greyaluminium_Normal_Visual.inst.cfg
weedo_x40_0.4_Material4Print_PETG_Brightred_Normal_Visual.inst.cfg
weedo_x40_0.4_Material4Print_PETG_Irongrey_Normal_Visual.inst.cfg
weedo_x40_0.4_Material4Print_PETG_Purered_Normal_Visual.inst.cfg
weedo_x40_0.4_Material4Print_PETG_Signalblue_Normal_Visual.inst.cfg
weedo_x40_0.4_Material4Print_PETG_Skyblue_Normal_Visual.inst.cfg
weedo_x40_0.4_Material4Print_PETG_Trafficyellow_Normal_Visual.inst.cfg
weedo_x40_0.4_Material4Print_PETG_Transparentblue_Normal_Visual.inst.cfg
weedo_x40_0.4_Material4Print_PETG_Transparentgreen_Normal_Visual.inst.cfg
weedo_x40_0.4_Material4Print_PETG_Transparentred_Normal_Visual.inst.cfg
weedo_x40_0.4_Material4Print_PETG_Transparentyellow_Normal_Visual.inst.cfg
weedo_x40_0.4_Material4Print_PETG_Transparent_Normal_Visual.inst.cfg
weedo_x40_0.4_Material4Print_PETG_White_Normal_Visual.inst.cfg
weedo_x40_0.4_Material4Print_PLA_Anthracite_Normal_Visual.inst.cfg
weedo_x40_0.4_Material4Print_PLA_Black_Normal_Visual.inst.cfg
weedo_x40_0.4_Material4Print_PLA_Blue_Normal_Visual.inst.cfg
weedo_x40_0.4_Material4Print_PLA_Brightorange_Normal_Visual.inst.cfg
weedo_x40_0.4_Material4Print_PLA_Brown_Normal_Visual.inst.cfg
weedo_x40_0.4_Material4Print_PLA_Dustygrey_Normal_Visual.inst.cfg
weedo_x40_0.4_Material4Print_PLA_Brightred_Normal_Visual.inst.cfg
weedo_x40_0.4_Material4Print_PLA_Green_Normal_Visual.inst.cfg
weedo_x40_0.4_Material4Print_PLA_Greyaluminium_Normal_Visual.inst.cfg
weedo_x40_0.4_Material4Print_PLA_Irongrey_Normal_Visual.inst.cfg
weedo_x40_0.4_Material4Print_PLA_Ivory_Normal_Visual.inst.cfg
weedo_x40_0.4_Material4Print_PLA_Magenta_Normal_Visual.inst.cfg
weedo_x40_0.4_Material4Print_PLA_Nature_Normal_Visual.inst.cfg
weedo_x40_0.4_Material4Print_PLA_Pearlblue_Normal_Visual.inst.cfg
weedo_x40_0.4_Material4Print_PLA_Purered_Normal_Visual.inst.cfg
weedo_x40_0.4_Material4Print_PLA_Purple_Normal_Visual.inst.cfg
weedo_x40_0.4_Material4Print_PLA_Signalblue_Normal_Visual.inst.cfg
weedo_x40_0.4_Material4Print_PLA_Trafficyellow_Normal_Visual.inst.cfg
weedo_x40_0.4_Material4Print_PLA_White_Normal_Visual.inst.cfg
weedo_x40_0.4_Material4Print_PLA_Yellowgreen_Normal_Visual.inst.cfg
weedo_x40_0.4_Material4Print_PLA_Perlgold_Normal_Visual.inst.cfg
weedo_x40_0.4_Verbatim_Durabio_Black_Normal_Visual.inst.cfg
weedo_x40_0.4_Verbatim_Durabio_Clear_Normal_Visual.inst.cfg
weedo_x40_0.4_Verbatim_Durabio_White_Normal_Visual.inst.cfg
weedo_x40_0.4_Verbatim_PETG_Redtransparent_Normal_Visual.inst.cfg
weedo_x40_0.4_Verbatim_PETG_Black_Normal_Visual.inst.cfg
weedo_x40_0.4_Verbatim_PETG_Bluetransparent_Normal_Visual.inst.cfg
weedo_x40_0.4_Verbatim_PETG_Blue_Normal_Visual.inst.cfg
weedo_x40_0.4_Verbatim_PETG_Clear_Normal_Visual.inst.cfg
weedo_x40_0.4_Verbatim_PETG_Greentransparent_Normal_Visual.inst.cfg
weedo_x40_0.4_Verbatim_PETG_Red_Normal_Visual.inst.cfg
weedo_x40_0.4_Verbatim_PETG_White_Normal_Visual.inst.cfg
weedo_x40_0.4_Verbatim_PLA_Black_Normal_Visual.inst.cfg
weedo_x40_0.4_Verbatim_PLA_Blue_Normal_Visual.inst.cfg
weedo_x40_0.4_Verbatim_PLA_Green_Normal_Visual.inst.cfg
weedo_x40_0.4_Verbatim_PLA_Grey_Normal_Visual.inst.cfg
weedo_x40_0.4_Verbatim_PLA_Natural_Normal_Visual.inst.cfg
weedo_x40_0.4_Verbatim_PLA_Red_Normal_Visual.inst.cfg
weedo_x40_0.4_Verbatim_PLA_White_Normal_Visual.inst.cfg
```
With the following overrides:
```
skin_monotonic=True
speed_print=40
top_bottom_thickness==wall_thickness
wall_thickness==wall_line_width * 3
```

This one shows the same issue as before, but also a different issue: There are multiple types of materials in here: PETG, PLA, Durabio and GreenTEC PLA. If different material profiles always or often share the same set of profiles, they should probably get the same material type.

This one is really difficult to properly solve though and we don't require it to be perfectly optimised. There is a human factor in there too: PLA and PETG probably should be two material types because they are very different chemically. But perhaps GreenTEC Pro or Durabio are blends of other chemicals and very similar to those chemicals. They require the same profiles, after all, so maybe they should be considered the same material type, even if the material manufacturer wants to present it as something completely unique and much better than anything their competition sells.

Solution
----
I hope you can at least remove the duplicates for the first issue. The second one is more difficult to solve and we can accept some slowdown in the start-up. But we try to reduce duplicates a bit because to load even just the metadata to see for which printers they apply already slows down the start-up for everyone, not just the Weedo users.

**Pull request closed without merging**